00-programming-fundamentals/08-intro-to-automated-tests.md

```diff
@@ -60,24 +66,273 @@ true
+
+```ruby
+# die_spec.rb
+
+require 'minitest/autorun'
+require 'minitest/reporters'
+require_relative 'die'
+
+describe "Testing Die Class" do
+  it "You can create an instance of Die" do
+    @die = Die.new
+
+    @die.class.must_equal Die
+  end
+end
+
+```
+
+Notice the line `@die.class.must_equal Die`.  Look at how english-friendly it is.  The idea with **BDD** is the that with our tests we are defining specifications or behaviors our code should follow.  
+
+Similar to Assertions Minitest adds a bunch of methods to all objects called *matchers*.  There are a large number of matchers, but you can usually get by with `must_equal`, `must_include`, `must_match` and `must_raise`.  You can see a full list of Minitest matchers at the bottom of this lesson along with examples.  
+
+### Exercise Writing Another Test
+
+So just like writing the assertion, now write another test for the roll method.  You should create an `it` block and use the `must_be` matcher.  
+
+### Check & Verify
+
+Check again with your partner when you are finished.  You can find a solution [here](https://github.com/AdaGold/die/blob/solution/specs/die_spec.rb)
+
+
+## Which Style Should I Use?
+
+So which style is better?  Functionally both do the same job, but the spec-style is more readable by non-technical people.  So we will, from this point on, use **Spec-style tests.**  However it is important for you to know that both exist and when you get to rails, a great deal more documentation exists for assert-style tests compared to spec-style.   
+
+## What Should I Test?
+
+More important that how you test your code is what you are testing.  If you're not testing the right things bugs can creep through your tests and into production code.  Many many many developers have trouble knowing what to test.   Here are some general guidelines. 
+
+*  Look at your code for branches (if statements and loops) and make sure that each branch of execution is tested.
+*  Test your methods with edge case values.
+	*  For example if your method takes an integer as a parameter you would test it with a positive number, a negative number and zero.
+	*  If your method took an array as a parameter you would test it with an empty array, a one element array and a large array.  
+*  Think about how someone might misuse your method, check for invalid or weird input.  If someone can break your code... they will.
+
+## Organizing Code
+
+As our projects get larger putting all our code in the same folder gets messy, especially with test files.  So we will be setting up projects like this:
+
+![Project Structure](images/project_folder.png "project structure")
+
+We will be placing our code in the **lib** folder, and our test specs in the **specs** folder.
```

Comment:
Mostly because Rake/organization is something that looked like we might want to skip and come back to at a later date.  Placing it at the end made it easier to skip if time, and too much material made it undesirable as an intro.

I can work it in earlier.